src/main/java/com/geoxp/geo/GeoXPShapeHelper.java

```diff
@@ -0,0 +1,261 @@
+   int[] bend = new int[16];
+   
+   for (int i = 1; i < 16; i++) {
+     long cell = ((long) i) << 60;
+     int idx = Arrays.binarySearch(acells, cell); 
+
+     if (idx >= 0) {
+       astart[i] = idx;
+     } else {
+       astart[i] = -idx - 1;
+       if (astart[i] >= acells.length || (acells[astart[i]] & 0xF000000000000000L) != cell) {
+         astart[i] = -1;
+       }
+     }
+            
+     idx = Arrays.binarySearch(bcells, cell);       
+     
+     if (idx >= 0) {
+       bstart[i] = idx;
+     } else {
+       bstart[i] = -idx - 1;
+       if (bstart[i] >= bcells.length || (bcells[bstart[i]] & 0xF000000000000000L) != cell) {
+         bstart[i] = -1;
+       }
+     }
+   }
+   
+   for (int i = 1; i < 16; i++) {
+     aend[i] = -1;
+     if (-1 != astart[i]) {
+       // Find the start of the next index
+       for (int j = i + 1; j < 16; j++) {
+         if (-1 != astart[j]) {
+           aend[i] = astart[j] - 1;
+           break;
+         }
+       }
+       if (-1 == aend[i]) {
+         if (i >= 8) {
+           for (int j = 1; j < 8; j++) {
+             if (-1 != astart[j]) {
+               aend[i] = astart[j] - 1;
+               break;
+             }
+           }
+         }
+         if (-1 == aend[i]) {
+           aend[i] = acells.length - 1;
+         }
+       }
+     }
+   }
+
+   for (int i = 1; i < 16; i++) {
+     bend[i] = -1;
+     if (-1 != bstart[i]) {
+       // Find the start of the next index
+       for (int j = i + 1; j < 16; j++) {
+         if (-1 != bstart[j]) {
+           bend[i] = bstart[j] - 1;
+           break;
+         }
+       }
+       if (-1 == bend[i]) {
+         if (i >= 8) {
+           for (int j = 1; j < 8; j++) {
+             if (-1 != bstart[j]) {
+               bend[i] = bstart[j] - 1;
+               break;
+             }
+           }
+         }
+         if (-1 == bend[i]) {
+           bend[i] = bcells.length - 1;
+         }
+       }
+     }
+   }
+
+   //
+   // Maintain a bitset of included cells
+   //
+   
+   BitSet froma = new BitSet(acells.length);
+   BitSet fromb = new BitSet(bcells.length);
+   
+   //
+   // Now iterate over the cells of both shapes by increasing resolution
+   //
+   
+   for (int res = 1; res < 16; res++) {
+     if (-1 != astart[res]) {
+       for (int i = astart[res]; i <= aend[res]; i++) {
+         // Ignore cells which are already included
+         if (froma.get(i)) {
+           continue;
+         }
+         long cell = acells[i] & RESOLUTION_MASKS[res];
+         // Iterate over the resolutions finer or equal to res in the other shape
+         for (int subres = res; subres < 16; subres++) {
+           if (-1 == bstart[subres]) {
+             continue;
+           }
+           // attempt to find the insertion point for the subcell 0 at subres
+           long subcell = cell | (((long) subres) << 60);
+           int idx = Arrays.binarySearch(bcells, bstart[subres], bend[subres] + 1, subcell);
```

Comment:
This signature of `Arrays.binarySearch` would require you to bump from Java 5 to Java 6.
As a result, you should update build.gradle with, for instance,
```
sourceCompatibility = 1.8
targetCompatibility = 1.8
```